x-pack/libbeat/management/unit.go

```diff
@@ -0,0 +1,400 @@
+	}
+}
+
+// getUnitState converts status.Status to client.UnitState
+func getStatus(s client.UnitState) status.Status {
+	switch s {
+	case client.UnitStateStarting:
+		return status.Starting
+	case client.UnitStateConfiguring:
+		return status.Configuring
+	case client.UnitStateHealthy:
+		return status.Running
+	case client.UnitStateDegraded:
+		return status.Degraded
+	case client.UnitStateFailed:
+		return status.Failed
+	case client.UnitStateStopping:
+		return status.Stopping
+	case client.UnitStateStopped:
+		return status.Stopped
+	default:
+		return status.Unknown
+	}
+}
+
+func getStreamStates(expected client.Expected) (map[string]unitState, []string) {
+
+	expectedCfg := expected.Config
+
+	if expectedCfg == nil {
+		return nil, nil
+	}
+
+	streamStates := make(map[string]unitState, len(expectedCfg.Streams))
+	streamIDs := make([]string, len(expectedCfg.Streams))
+
+	for idx, stream := range expectedCfg.Streams {
+		streamState := unitState{
+			state: status.Unknown,
+			msg:   "",
+		}
+
+		if id := stream.GetId(); id != "" {
+			streamIDs[idx] = id
+			streamStates[id] = streamState
+			continue
+		}
+
+		if cfgName := expectedCfg.GetName(); cfgName != "" {
+			id := fmt.Sprintf("%s.[%d]", cfgName, idx)
+			streamIDs[idx] = id
+			streamStates[id] = streamState
+			continue
+		}
+
+		id := fmt.Sprintf("%s.[%d]", expectedCfg.GetId(), idx)
+		streamIDs[idx] = id
+		streamStates[id] = streamState
+	}
+
+	return streamStates, streamIDs
+}
+
+// newAgentUnit creates a new agentUnit. In case the supplied client.Unit is of type
+// client.UnitTypeInput it initializes the streamStates with a unitState.Unknown
+func newAgentUnit(cu *client.Unit, log *logp.Logger) *agentUnit {
+	var (
+		streamStates map[string]unitState
+		streamIDs    []string
+	)
+
+	if cu.Type() == client.UnitTypeInput {
+		streamStates, streamIDs = getStreamStates(cu.Expected())
+	}
+
+	return &agentUnit{
+		clientUnit:   cu,
+		logger:       log,
+		streamIDs:    streamIDs,
+		streamStates: streamStates,
+	}
+}
+
+// RegisterAction registers action handler for this unit.
+func (u *agentUnit) RegisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.RegisterAction(action)
+}
+
+// UnregisterAction unregisters action handler with the client.
+func (u *agentUnit) UnregisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.UnregisterAction(action)
+}
+
+func (u *agentUnit) Expected() client.Expected {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return client.Expected{}
+	}
+
+	return u.clientUnit.Expected()
+}
+
+func (u *agentUnit) ID() string {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return ""
+	}
+
+	return u.clientUnit.ID()
+}
+
+// calcState calculates the current state of the unit.
+func (u *agentUnit) calcState() (status.Status, string) {
```

Comment:
This method should be unit tested.

Reply:
UpdateState may be what you actually want to test now that I've read that far down.

Reply:
added the requested unit-test, feel free to un-resolve if you spot anything